### DISCUSSION TITLE
How to generate *.lib file?

### DISCUSSION THREAD

**Discussion by mhughson** [cc65/cc65#1535] [Q&A] [answered]
Is it possible to compile a program in CC65 to a statically linked library file? I have found information in the docs regarding linking _in_ a lib file, but I can't seem to find anything about building one.

If it is relevant, I am targeting the NES.

For context, I want to distribute a tool for my game, which includes compiling a ROM, but I don't want to distribute all the source code, so my thought was that a majority of the code could be included as a *.lib file to help obfuscate the code. Perhaps there are other ways.

**Comment by mrdudz** [comment]
see ar65 docs... simply said you compile all your code into .o (object) files, and produce a library with ar65

**Reply by mhughson** [reply]
Ok cool. I wasn't aware of AR65, so I'll dig into that! Thanks!

**Comment by mhughson** [comment, marked as answer]
Thanks to @mrdudz for pointing me in the right direction. Just wanted to circle back and share what I ended up with, in case anyone else out there is attempting the same thing.

My link command originally looked like this:

`ld65 -C MMC1_128_128.cfg --dbgfile %name%.dbg -o %name%.nes crt0.o %name%.o mmc1\bank_helpers.o PRG0.o PRG1.o PRG2.o PRG3.o PRG4.o PRG5.o nes.lib -Ln labels.txt`

The equivalent with a static library introduced is this:

```
REM Put PRG0-5.o into a static library, "game.lib"
ar65 r game.lib PRG0.o PRG1.o PRG2.o PRG3.o PRG4.o PRG5.o
ld65 -C MMC1_128_128.cfg --dbgfile %name%.dbg -o %name%.nes crt0.o %name%.o mmc1\bank_helpers.o game.lib nes.lib -Ln labels.txt
```

Documentation is here: https://cc65.github.io/doc/ar65.html

**Reply by oliverschmidt** [reply]
If your crt0.s exports `__STARTUP__`, you can put it into your lib too - see e.g. https://github.com/cc65/cc65/blob/09862e7ce908d7915e6341f626f2930e0cc62118/libsrc/nes/crt0.s#L9

**Reply by mhughson** [reply]
> If your crt0.s exports `__STARTUP__`, you can put it into your lib too - see e.g.
> 
> https://github.com/cc65/cc65/blob/09862e7ce908d7915e6341f626f2930e0cc62118/libsrc/nes/crt0.s#L9

Can you explain in more detail? 

My crt0.s _does_ export __STARTUP__. but when you say "you can put _it_ into your lib too", what is "it"? What would change about my build script posted about (the ar65 and ld65 lines)?

Also, what problems does this other way of doing things address (just so I can understand why I would do one over the other)?

I am fairly uneducated with the CC65 pipeline, so apologies if these questions have obvious answers. I appreciate your time!

**Reply by oliverschmidt** [reply]
'it' refers to your crt0.s (or rather crt0.o).

I guess your interest in creating a static library comes from the desire to make ld65 cmdlines shorter. By adding your crt0.o to the static library (just as you did with PRGx.o) you can omit it from the ld65 cmdline (just as you did with PRGx.o).

**Reply by mhughson** [reply]
Ah, OK, I see what you mean. Thanks for clarifying! I am intentionally not including all object files in the lib. 

My interest is actually just to be able to allow people "mod" my game, without making the full source public. So my lib will contain the _core_ logic of the game, while users can add their own "data", and compile it all together. The length of the command line is not a concern for me.
